### PR TITLE
Updating the project pages to meet requirements

### PR DESCRIPTION
**Pull request by codyrancher**
rancher/dashboard#567

Screen Shot 2020-06-18 at 1 19 25 PM
Screen Shot 2020-06-18 at 1 19 21 PM
Screen Shot 2020-06-18 at 1 19 14 PM

### CODE REVIEW

File: components/BreadCrumbs.vue [hunk 1]
```diff
@@ -20,9 +28,15 @@ export default {
       if (crumbLocations[i - 1]) {
         nextName = ( `${ crumbLocations[i - 1].name }-${ piece }`);
       }
+      const nextParams = this.paramsFor(nextName, params);
+
+      if (nextName === 'c-cluster-resource' && params.resource === EXTERNAL.PROJECT) {
+        nextParams.resource = NAMESPACE;
+      }
```
Comment (vincent99) [inline components/BreadCrumbs.vue:35]: I don't get what this is accomplishing?  Your screenshots show editing a project but the breadcrumb saying namespace; and the type missing entirely in `Edit <type>: <id>` header.

Comment (codyrancher) [reply components/BreadCrumbs.vue:35]: I don't think it's necessary. I suspect it was exploratory code while modifying this so that it shows namespace when in a project detail page since we don't have a project list page.

The type missing is completely unrelated to this. That was caused by masthead looking up the schema for project in the wrong module. I also made that change.

File: components/form/NameNsDescription.vue [hunk 2]
```diff
@@ -63,10 +63,12 @@ export default {
       metadata.namespace = this.$store.getters['defaultNamespace'];
     }
     const description = metadata.annotations?.[DESCRIPTION];
+    const isExternal = Object.values(EXTERNAL).includes(this.value.type);
+    const name = isExternal ? this.value.spec.displayName : metadata.name;
```
[the remark this answers is not in the document]
Comment (codyrancher) [reply components/form/NameNsDescription.vue:67]: Switched but I still needed that check later on in the name watcher.

File: edit/project/index.vue [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        }
+      });
+    }
+  }
+};
+</script>
+
+<template>
+  <div class="project">
+    <NameNsDescription
```
[the remark this answers is not in the document]
Comment (codyrancher) [reply edit/project/index.vue:140]: added and handled spec.displayName.

File: assets/translations/en-us.yaml [hunk 4]
```diff
@@ -769,3 +769,37 @@ model:
     displayKind:
       generic: Service
       loadBalancer: L4 Balancer
+projectPage:
+  namespaces: Namespaces
+  tabs:
+    resourceQuota:
+      placeholders:
+        2000: "e.g. 2000"
```
[the remark this answers is not in the document]
Comment (codyrancher) [reply assets/translations/en-us.yaml:777]: Switched.

File: components/BreadCrumbs.vue (hunk 1 above)
Comment (codyrancher) [reply components/BreadCrumbs.vue:35]: Actually it wasn't exploratory. It was to change the link to link to namespace instead of project. The other change in here just changes what's displayed not what's linked.

**Comment by vincent99** [comment]
I don't mean to pick on this particular PR but there are multiple obvious things wrong here, some of which are literally in the screenshots provided, and this has become a recurring problem.  There needs to be more thorough testing going on and more attention paid to details (for **everyone**) before submitting PRs to make sure the feature:

  1. Works
  1. Makes sense
  1. Looks good
  1. Is consistent with the rest of the product.

- The header saying Namespaces for a project is completely misleading.  Nothing tells me I'm looking at a project. (yes I know the parent page is currently labeled Namespaces, but that doesn't make this ok).
image

- Project detail has no detail top with create date/etc like every other resource (and created is for some reason not a standard part of that control, even though every resource has one).  There's a chasm of whitespace shown there instead.
image

- Project description isn't shown anywhere.  Again I don't know why this isn't a standard part of the header.

- Project YAML view is broken, shows an empty box.
image

- Project detail has a cancel button for no reason.  This has been in multiple other places as well so this isn't unique to this page but clearly nobody's ever going to fix it without being pointed out.  Cancel from the broken YAML view explodes.

- Container resource limit tab shows all n/a's instead of one message that says there are none like other places are starting to do
image
image

- The info text for limit isn't in a banner
image


- Resource Quota table is missing noRows content
image

- The project dropdown in create NS is broken (probably not by this PR, but should have seen it while testing it):
image

- Is this really how we want to display a list of potentially many namespaces?
image

- Editing a project causes an API call even though they're all loaded (because the id has been [prefixed](https://github.com/rancher/dashboard/blob/master/components/ResourceDetail/index.vue#L95-L97) and doesn't match), and then ends up loading a new model, which gives you duplicate entries in the store that show up in create NS.  The namespaces in clusterExternal are one per managed cluster or project. 
image

- Namespaces belong to zero-or-one projects.  The namespace picker in Edit Project shouldn't offer namespaces that are already in some other project (or already in another entry of the editor list, really).  This add/dropdown/remove style list is not a particularly appropriate control for this usage.
image

- If you turn off namespace grouping by project, the projects column isn't linked so the project details are unreachable.
image

- I can edit projects but not create them?

- "Container default resource limits" on namespaces are a Rancher feature and should only be shown if in `isRancher` mode.

**Comment by codyrancher** [comment]
@vincent99 it looks like most of the concerns you listed are actually about the detail page which is still TBD (though it should mostly look like read version of the edit page) https://github.com/rancher/dashboard/issues/567 or the namespace page which I didn't actually test or modify much since the changes were completed elsewhere by someone else https://github.com/rancher/dashboard/issues/568. I will attempt to address the others via comment. I probably shouldn't have posted the detail page screenshots and probably should have linked the individual project issue instead.

> The header saying Namespaces for a project is completely misleading. Nothing tells me I'm looking at a project. (yes I know the parent page is currently labeled Namespaces, but that doesn't make this ok).

It seems that maybe it should be renamed to Projects/Namespaces like Rancher has. I didn't notice a specific call out for this and was told a couple times that we're trying to deemphasize projects so I chose to have the project breadcrumb lead back to the namespace route. Given the  context I had this choice seemed the least misleading as it brought you back to the page you would use to get to the detail page. What do you see as a good solution for this?

> Project detail has no detail top with create date/etc like every other resource (and created is for some reason not a standard part of that control, even though every resource has one). There's a chasm of whitespace shown there instead.

That's definitely my fault. I'll go ahead and fix it.

> Project YAML view is broken, shows an empty box.

Is there a way to currently get the yaml from clusterExternal resources? The yaml link present on the steve resource doesn't appear to be present for clusterExternal resources.

> Editing a project causes an API call even though they're all loaded (because the id has been prefixed and doesn't match), and then ends up loading a new model, which gives you duplicate entries in the store that show up in create NS. The namespaces in clusterExternal are one per managed cluster or project.

I'll dig into this one more but I don't fully understand it at the moment. It seems like ultimately I should work on ensuring the extraneous call isn't made. It seems like conceptually maybe all resources should be stored and requested using fqid but maybe there's a reason that can't be done.

> Namespaces belong to zero-or-one projects. The namespace picker in Edit Project shouldn't offer namespaces that are already in some other project (or already in another entry of the editor list, really). This add/dropdown/remove style list is not a particularly appropriate control for this usage.

To me it seems like the control would be appropriate as long as the options were filtered appropriately. Would you agree? If not, do you have a suggestion?

> If you turn off namespace grouping by project, the projects column isn't linked so the project details are unreachable.

I'll fix this, good catch.

> I can edit projects but not create them?

This is what I was told by @deniseschannon 
_we will not have a specific "create project" page but we will still need the ability to edit/view projects. we will incorporating being able to simply add a project into the namespace creation workflow and if you want to set thse other fancier things for project, you woul d need to edit the project to enable those features. we're trying to start downplaying project usage for new users_

> "Container default resource limits" on namespaces are a Rancher feature and should only be shown if in isRancher mode.

I'll go ahead and add this condition.

**Comment by vincent99** [comment]
> I'll dig into this one more but I don't fully understand it at the moment. It seems like ultimately I should work on ensuring the extraneous call isn't made. It seems like conceptually maybe all resources should be stored and requested using fqid but maybe there's a reason that can't be done.

TL;DR you shouldn't be adding the namespace to the fqid in the external case... `if ( !isExternalType && schema?.attributes?.namespaced && namespace ) {`.  That will have the side-effect of matching the entry already in store and not making a second request or adding a 2nd copy to the store.

"fqid"s are sort of a cheat to inject a namespace collection level between the type collection and individual resources into an API system that doesn't really know about them.  For namespaced resources the type is say `configmap`, and the id API returns is `<namespace>/<name>`.  When that is put together into a url it becomes `.../v1/configmaps/<namespace>/<name>` to get an individual resource, which works.  If you take off the name you can get just the configmaps for that ns (though the UI doesn't use this), and if you go one more level you get all the configmaps.

But the clusterExternal API is a special filtered view of the resources which tells you "for the given cluster Foo, what resources exist in the management cluster but are for Foo"?  Those resources are stored in mgmt cluster in a namepace for each cluster (e.g. if foo is `c-a1934f` then there's a namespace in mgmt called `c-a1934f` in it).

These views are mounted "inside" the cluster (`/v1/...clusters/c-a1934f/projects`) so the namespace to be looked for is already known (it's the cluster id) and is therefore made invisible to you the client and not part of the resource ID.  So it's just ``/v1/...clusters/c-a1934f/projects` for the list and `/v1/...clusters/c-a1934f/projects/p-8384a` for a single project inside that.


> To me it seems like the control would be appropriate as long as the options were filtered appropriately. Would you agree? If not, do you have a suggestion?

I do not; if you have N independent dropdowns and only want to show valid unique options in each:
  - The available options for every dropdown depends on selected value of all the N-1 other dropdowns.
  - Every one has to be recomputed each time any one changes.
  - Each dropdown the user clicks has a slightly different list of values in it, for what will not be an obvious reason.
  - Changing a value in one dropdown sometimes results in one action (add to this ns) and sometimes two (remove from old ns, add to new).  Changing the existing value in an existing dropwdown doesn't sound as dangerous as clicking "remove" but has the same effect.

So I would start the bidding at: don't have management of the n:1 relationship in project edit at all.  This is what 2.x does.  You can manage what namespaces belong to a project with bulk actions in the list page (or potentially drag & drop), and you can move a single namespace to a different project from the namespace details, but not manage many namespaces in project details.

If we do still want n:1 editing on project detail, I would present the dropdown of choices only once.  The obvious choice seems like a `<[v-]Select :multiple="true" ...>`.  The namespaces that already belong to a project (!= this one) could show as disabled options, or be hidden entirely.  Picking up a NS that has no projectId is relatively harmless, but I definitely don't think you should be able to move a NS from [an assigned project] to [this project] in this screen.

Alternately you could do lots of checkboxes `<label v-for="ns in namespaces"><input type="checkbox" value="ns.id"> {{ns.nameDisplay}}</label>`

Or show static text values for the currently selected rows with a remove button on each and a single dropdown + add button to add one more (not that we have anything like this currently):

```
[ select a namespace | v ] [ Add ]

current ns 1  [remove]
current ns 2  [remove]
current ns 3 [remove]
```
